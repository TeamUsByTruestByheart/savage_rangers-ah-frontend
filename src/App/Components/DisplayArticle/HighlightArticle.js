import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

export class HighlightArticle extends Component {
  state = {
    highlightedText: '',
    anchorKey: '',
    startIndex: 0,
    lastIndex: 0,
    selectionRectangle: null,
    isModalShown: false,
    comment: '',
    nodeId: null,
    highlights: null,
    originalHtml: null
  };

  componentDidMount = () => {
    window.addEventListener('mouseup', this.checkHighlightedText);
    window.addEventListener('click', this.checkHighlightedText);
    window.addEventListener('resize', this.checkHighlightedText);
  };

  getSelectionRange = (windowSelection) => {
    const selectionRange = windowSelection.toString() && windowSelection.getRangeAt(0);
    const selectionRectangle = selectionRange && selectionRange.getBoundingClientRect();
    return { selectionRange, selectionRectangle };
  };

  checkHighlightedText = () => {
    const { isModalShown } = this.state;
    if (isModalShown) return false;
    const windowSelection = window.getSelection();
    const { selectionRectangle } = this.getSelectionRange(windowSelection);
    const startIndex = Math.min(windowSelection.anchorOffset, windowSelection.focusOffset);
    const lastIndex = Math.max(windowSelection.anchorOffset, windowSelection.focusOffset);
    const highlightedText = windowSelection.toString();
    const nodeId = window.getSelection().anchorNode.parentNode.id;

    return this.setState(prevState => ({
      ...prevState,
      startIndex: highlightedText ? startIndex : prevState.startIndex,
      lastIndex: highlightedText ? lastIndex : prevState.lastIndex,
      highlightedText: highlightedText || prevState.highlightedText,
      selectionRectangle,
      nodeId: nodeId || prevState.nodeId
    }));
  };

  openCommentModal = () => {
    this.setState(prevState => ({
      ...prevState,
      isModalShown: true,
      comment: ''
    }));
  };

  closeCommentModal = () => {
    this.setState(prevState => ({
      ...prevState,
      isModalShown: false,
      selectionRectangle: null,
      comment: ''
    }));
  };

  handleChange = ({ target }) => {
    const input = { [target.name]: target.value };

    this.setState(prevState => ({
      ...prevState,
      ...input
    }));
  };

  handleSubmit = (e) => {
    e.preventDefault();
    const {
      highlightedText: text, startIndex, lastIndex, comment, nodeId
    } = this.state;
    const { postHighlight, slug } = this.props;

    const formData = {
      text,
      startIndex,
      lastIndex,
      comment,
      nodeId
    };

    postHighlight(formData, slug);
  };

  render() {
    const {
      selectionRectangle,
      isModalShown,
      highlightedText,
      comment,
      startIndex,
      lastIndex,
      nodeId
    } = this.state;

    const { highlightDetails } = this.props;
    return (
      <div>
        {selectionRectangle && !isModalShown ? (
          <div
            style={{
              position: 'absolute',
              top: `${selectionRectangle.top + 30 + window.scrollY}px`,
              left: `${selectionRectangle.left}px`,
              zIndex: 12345
            }}
          >
            <button
              type="button"
              className="btn btn-warning"
              onClick={this.openCommentModal}
              data-toggle="modal"
              data-target="#highlightCommentModal"
            >
              <i className="fas fa-comment" />
            </button>
          </div>
        ) : (
          ''
        )}

        <div
          className="modal"
          id="highlightCommentModal"
          tabIndex="-1"
          role="dialog"
          aria-labelledby="highlightCommentModalLabel"
          aria-hidden="true"
        >
          <div className="modal-dialog" role="document">
            <div className="modal-content">
              <div className="modal-header">
                <h5 className="modal-title" id="highlightCommentModalLabel">
                  Comment Highlight
                </h5>
                <button
                  type="button"
                  className="close"
                  data-dismiss="modal"
                  aria-label="Close"
                  onClick={this.closeCommentModal}
                >
                  <span aria-hidden="true">&times;</span>
                </button>
              </div>
              <form id="comment-highlight-form" onSubmit={this.handleSubmit}>
                <div className="modal-body">
                  <textarea
                    name="comment"
                    row="4"
                    className="form-control"
                    value={comment}
                    onChange={this.handleChange}
                  />
                  <hr />
                </div>
                <div className="modal-footer">
                  <button
                    type="button"
                    className="btn btn-secondary"
                    data-dismiss="modal"
                    onClick={this.closeCommentModal}
                  >
                    Close
                  </button>
                  <button type="submit" className="btn btn-primary">
                    Save
                  </button>
                </div>
              </form>
            </div>
          </div>
        </div>
        {/* Higlight details */}
        <div
          className="modal"
          id="highlightDetailsModal"
          tabIndex="-1"
          role="dialog"
          aria-labelledby="highlightDetailsModalLabel"
          aria-hidden="true"
        >
          <div className="modal-dialog" role="document">
            <div className="modal-content">
              <div className="modal-header">
                <h5 className="modal-title" id="highlightDetailsModalLabel">
                  Highlight details
                </h5>
                <button type="button" className="close" data-dismiss="modal" aria-label="Close">
                  <span aria-hidden="true">&times;</span>
                </button>
              </div>
              <form onSubmit={this.handleSubmit}>
                <div id="highlightDetailsModalBody" className="modal-body">
                  Hello
                </div>
                <div className="modal-footer">
                  <button type="button" className="btn btn-secondary" data-dismiss="modal">
                    Close
                  </button>
                </div>
              </form>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

HighlightArticle.propTypes = { postHighlight: PropTypes.func };

export default connect(null,
  null)(HighlightArticle);
