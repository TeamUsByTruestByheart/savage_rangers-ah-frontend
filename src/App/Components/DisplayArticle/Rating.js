/* eslint-disable jsx-a11y/click-events-have-key-events */
/* eslint-disable jsx-a11y/no-static-element-interactions */
import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import isEmpty from 'lodash/isEmpty';
import TriangularPopup from '../Common/TriangularPopup';
import actions from '../../../Redux/Actions/articleRatings';

const { showUsersForRating } = actions;

export class Rating extends Component {
  static propTypes = {
    rating: PropTypes.number,
    percentage: PropTypes.number,
    stats: PropTypes.array,
    areUsersShown: PropTypes.bool,
    showUsersForRating: PropTypes.func
  };

  constructor(props) {
    super(props);
    this.state = { page: 1, showUsers: false };
  }

  showOrHide = () => {
    const { props: { areUsersShown, showUsersForRating: showUsers_ } } = this;
    if (!areUsersShown) {
      const { state: { showUsers } } = this;
      this.setState({ showUsers: !showUsers });
      showUsers_(!areUsersShown);
    }
  };

  hide = () => {
    const { props: { areUsersShown, showUsersForRating: showUsers_ } } = this;
    this.setState({ showUsers: false });
    showUsers_(!areUsersShown);
  };

  gotoNextPage = () => {
    const { state: { page } } = this;
    this.setState({ page: page + 1 });
  };

  gotoPrevPage = () => {
    const { state: { page } } = this;
    this.setState({ page: page - 1 });
  };

  render() {
    const {
      props: { rating, percentage, stats },
      state: { page, showUsers }
    } = this;
    const pages = parseFloat(stats[rating - 1].users[0]);
    const users = stats[rating - 1].users[page];
    return (
      <React.Fragment>
        <div className="rating" onClick={this.showOrHide}>
          <div className="rating__number">{rating}</div>
          <div className="rating__star">
            <i className="fas fa-star" />
          </div>
          <div className="rating__fill">
            <div className={`fill fill--${rating}`} style={{ width: `${percentage}%` }}>
              _
            </div>
          </div>
        </div>
        <div className="rating__users" style={{ display: showUsers ? 'block' : 'none' }}>
          <i className="fas fa-times-circle rating-close-button" onClick={this.hide} />
          <TriangularPopup direction="left" className="custom-triangle-rating" />
          {users
            && users.map(user => (
              <div key={user.id} className="rating__users__user">
                <div className="rating__users__user--username">
                  <a href={`/profile/${user.username}`}>{user.username}</a>
                </div>
                <div className="rating__users__user--rating">{rating}</div>
                <div className="rating__users__user--star">
                  <i className="fas fa-star" />
                </div>
              </div>
            ))}
          {isEmpty(users) ? (
            <div className="no-users">
              No users for this rating
              {' '}
              <i className="fas fa-user-times" />
            </div>
          ) : (
            ''
          )}
          {pages > 1 ? (
            <div className="rating__users__pagination">
              {page !== 1 ? (
                <div className="rating__users__pagination--left" onClick={this.gotoPrevPage}>
                  <i className="fas fa-arrow-left" />
                </div>
              ) : (
                ''
              )}
              <div className="rating__users__pagination--mid rounded-circle">{page}</div>
              {page !== pages ? (
                <div className="rating__users__pagination--right" onClick={this.gotoNextPage}>
                  <i className="fas fa-arrow-right" />
                </div>
              ) : (
                ''
              )}
            </div>
          ) : (
            ''
          )}
        </div>
      </React.Fragment>
    );
  }
}

export const mapStateToProps = ({ articleRatings }) => ({ areUsersShown: articleRatings.areShown });

export default connect(mapStateToProps,
  { showUsersForRating })(Rating);
