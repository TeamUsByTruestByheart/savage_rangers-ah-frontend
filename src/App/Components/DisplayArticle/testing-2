const ht = document.querySelector('#node-6233').innerHTML.substring(311, 320);
const sht = `<span style="color: red" onclick="alert('hello')">${ht}</span>`;
const updatedHtml = document.querySelector('#node-6233').innerHTML.replace(ht, sht);
document.querySelector('#node-6235').innerHTML = updatedHtml
