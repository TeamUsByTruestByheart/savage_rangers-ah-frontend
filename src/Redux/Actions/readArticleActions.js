import types from './index';
import axios from '../../configs/axios';

const { READ_ARTICLE, GET_ARTICLE_TAGS } = types;

export default {
  readArticle: slug => (dispatch) => {
    axios
      .get(`https://authors-heaven.herokuapp.com/api/articles/${slug}`)
      .then((res) => {
        dispatch({
          type: READ_ARTICLE,
          payload: res.data.article
        });
      });
  },
  getTags: slug => (dispatch) => {
    axios
      .get(`https://authors-heaven.herokuapp.com/api/articles/${slug}/tags`)
      .then((res) => {
        dispatch({
          type: GET_ARTICLE_TAGS,
          payload: res.data
        });
      });
  }
};
