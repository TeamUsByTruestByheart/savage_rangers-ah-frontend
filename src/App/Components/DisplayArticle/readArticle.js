/* eslint-disable max-len */
/* eslint-disable react/destructuring-assignment */
import React, { Component } from 'react';
import { connect } from 'react-redux';
import propTypes from 'prop-types';
import moment from 'moment';
import ReactImageFallback from 'react-image-fallback';
import Popular from '../Popular/popular';
import Actions from '../../../Redux/Actions/readArticleActions';
import Navbar from '../Common/NavProfile/navbar';
import Footer from '../Common/Footer';

const { readArticle, getTags } = Actions;

export class ReadArticle extends Component {
  state = { isBookmarked: false, tags: [] }

  componentWillMount() {
    const { readArticle: readArticles, getTags: getTag } = this.props;
    const { match: { params: { slug } } } = this.props;
    readArticles(slug);
    getTag(slug);
  }


  componentWillReceiveProps(nextProps) {
    if (nextProps.article && nextProps.tags) {
      this.setState({
        slug: nextProps.article.slug,
        category: nextProps.article.Category.name,
        title: nextProps.article.title,
        body: nextProps.article.body,
        readTime: nextProps.article.readTime,
        createdAt: nextProps.article.createdAt,
        coverImage: nextProps.article.coverImage,
        firstName: nextProps.article.User.firstName,
        lastName: nextProps.article.User.lastName,
        profileImage: nextProps.article.User.profileImage,
        tags: nextProps.tags.data
      });
    }
  }

  componentWillUpdate(nextProps) {
    const { readArticle: readArticles, getTags: getTag } = this.props;
    const { params: { slug } } = nextProps.match;
    if (nextProps.match.params.slug !== this.state.slug) {
      readArticles(slug);
      getTag(slug);
    }
  }

  render() {
    const {
      category, title, body, readTime, tags,
      isBookmarked, createdAt, coverImage,
      firstName, lastName, profileImage
    } = this.state;
    const formattedDate = moment(createdAt).format('Do MMM YYYY');
    return (
      <React.Fragment>
        <Navbar />
        <div className="page__label">
          <p className="page__label--text">Articles</p>
          <div className="category-p">
            <p className="category-p__label">CATEGORY</p>
          </div>
          <div className="category-t">
            <p className="category-t__label">{category}</p>
          </div>
        </div>
        <div className="container article">
          <div className="row">
            <div className="col-sm-1 article__icons">
              <button type="button" className="article__rating">
                <i className="fas fa-star" id="star" />
              </button>
              <i className={isBookmarked ? 'fas fa-bookmark' : 'far fa-bookmark'} id="bookmark" />
            </div>
            <div className="col-sm-9">
              <div className="article__data">
                <h2 className="article__description">{title}</h2>
                <div className="article__profileDetails">
                  <ReactImageFallback className="article__imgProfile" src={profileImage} fallbackImage="https://cdn.pixabay.com/photo/2015/10/05/22/37/blank-profile-picture-973460_960_720.png" alt="profilePic" title="profile" />
                  <p to="/" className="article__nameDetails">{`${firstName} ${lastName}`}</p>
                  <p to="/" className="article__dateDetails">{formattedDate}</p>
                  <p to="/" className="article__minDetails">
                    {readTime}
                    {' '}
                    min
                  </p>
                </div>
                <ReactImageFallback style={{ width: '600px', height: '500px' }} id="coverImage" src={coverImage} fallbackImage="https://ielektro.es/wp-content/uploads/2017/04/ventajas-comprar-LED.jpg" alt="coverImage" />
                <p className="article__body">{body}</p>
                {
                  tags.map(item => (
                    <div key={item.id} className="article__tags">
                      {item.name}
                    </div>
                  ))
                }
              </div>
            </div>
            <div className="col-sm-2">
              <Popular />
            </div>
          </div>
        </div>
        <Footer />
      </React.Fragment>
    );
  }
}
ReadArticle.propTypes = { article: propTypes.object, tags: propTypes.object };
export const mapStateToProps = state => ({ article: state.readArticle.article, tags: state.readArticle.tags });
export default connect(mapStateToProps, { readArticle, getTags })(ReadArticle);
