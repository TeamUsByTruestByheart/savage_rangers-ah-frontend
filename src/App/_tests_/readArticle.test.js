import React from 'react';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import mockAxios from 'axios';
import promiseMiddleware from 'redux-promise-middleware';
import { shallow } from '../../enzyme';
import readArticleActions from '../../Redux/Actions/readArticleActions';
import { ReadArticle, mapStateToProps } from '../Components/DisplayArticle/readArticle';
import Readreducer from '../../Redux/Reducers/readArticleReducer';

const { readArticle, getTags } = readArticleActions;
const middlewares = [thunk, promiseMiddleware];
const mockStore = configureStore(middlewares);

const initialState = { Article: {} };
const store = mockStore(initialState);

const payload = {
  article: {
    id: 6,
    title: 'What you need to know to become a great software engineer in 2020',
    description: '10 Tips to become a great software engineer by 2020.',
    body: 'I get quite a few emails that basically say',
    slug: 'what-you-need-to-know-to-become-a-great-software-engineer-in-2020-nfovo9fghng',
    readTime: 6,
    coverImage: 'https://res.cloudinary.com/al-tech/image/upload/v1565787533/rvkmbae3pcqhgysxoqnk.jpg',
    author: 42,
    category: 4,
    isBlocked: false,
    status: 'published',
    createdAt: '2019-08-14T12:58:54.185Z',
    updatedAt: '2019-08-16T09:19:39.885Z',
    deletedAt: null,
    Category: { name: 'TECHNOLOGY' },
    User: {
      username: 'MCFrank16',
      firstName: 'Frank',
      lastName: 'Mutabazi',
      profileImage: 'https://res.cloudinary.com/al-tech/image/upload/v1565805494/nenxg6yetjfgbk96iq8l.jpg'
    }
  }
};

mapStateToProps({ readArticle: {} });
const match = { params: { slug: 'slug' } };
const props = {
  readArticle: jest.fn(), getTags: jest.fn(), match, tags: { data: [{ id: 1, name: 'TECH' }] }, article: { Category: { name: 'Tech' }, User: { firstName: 'Frank' } }
};
const display = shallow(<ReadArticle {...props} />);
describe('Display', () => {
  it('renders the Display component', () => {
    expect(display).toBeDefined();
    expect(display.find('div').exists()).toBe(true);
    expect(display).toMatchSnapshot();
  });

  it('should receive the props', () => {
    display.setState({ article: {}, tags: [], isBookmarked: true });
    display.setProps(props);
    expect(display.instance().props).toEqual(props);
    expect(display.instance().state).toEqual({
      article: {}, body: undefined, category: 'Tech', coverImage: undefined, createdAt: undefined, firstName: 'Frank', isBookmarked: true, lastName: undefined, profileImage: undefined, readTime: undefined, slug: undefined, tags: [{ id: 1, name: 'TECH' }], title: undefined
    });
  });
});

describe('test the reducers', () => {
  const tagPayload = {
    tags: {
      status: 200,
      data: [
        {
          id: 4,
          name: 'Software',
          createdAt: '2019-08-14T12:56:55.150Z',
          updatedAt: '2019-08-14T12:56:55.150Z'
        },
        {
          id: 6,
          name: 'Growing',
          createdAt: '2019-08-14T12:56:55.152Z',
          updatedAt: '2019-08-14T12:56:55.152Z'
        },
        {
          id: 5,
          name: 'Engineering',
          createdAt: '2019-08-14T12:56:55.151Z',
          updatedAt: '2019-08-14T12:56:55.151Z'
        }
      ]
    }
  };
  it('should handle the read article action result', () => {
    expect(Readreducer(initialState, {
      type: 'READ_ARTICLE',
      payload
    })).toEqual({
      Article: {},
      article: { ...payload }
    });
  });

  it('should handle the get tags by article action result', () => {
    expect(Readreducer(initialState, {
      type: 'GET_ARTICLE_TAGS',
      payload: tagPayload
    })).toEqual({
      Article: {},
      tags: { ...tagPayload }
    });
  });
});

describe('testing the actions', () => {
  it('should check the read article action on failure', () => {
    const dataTest = { body: 'this is not awesome' };
    mockAxios.get = jest.fn(() => Promise.reject({ response: { data: { errors: {} } } }));
    store.dispatch(readArticle(dataTest));
    expect(store.getActions()[0]).toBe(undefined);
  });

  it('should check the get article tags action on failure', () => {
    const dataTest = { body: 'this is awesome' };
    mockAxios.get = jest.fn(() => Promise.reject({ response: { data: { errors: {} } } }));
    store.dispatch(getTags(dataTest));
    expect(store.getActions()[0].type).toEqual('CATCH_ERROR');
  });
});
