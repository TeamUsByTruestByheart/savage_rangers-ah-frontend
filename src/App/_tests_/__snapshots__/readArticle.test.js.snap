// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Display renders the Display component 1`] = `ShallowWrapper {}`;
